### PR TITLE
Manage User Addresses

### PR DESCRIPTION
**Pull request by anim-13**
Implemented Manage User Address task. In this task i had created a new model called Address and implemented following features :
1. Create a new address.
2. Show all the user addresses. 
3. Edit the user address.
4. Delete the user address.

Trello Link : https://trello.com/c/52mRR9uA

Screenshot from 2021-03-16 14-33-15
Screenshot from 2021-03-16 15-15-34

### CODE REVIEW

File: app/controllers/addresses_controller.rb [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    @address = current_user.addresses
+  end
+
+  def new
+    @address = @user.addresses.new
+  end
+
+  def create  
+    @address = @user.addresses.create(address_params)
+    if @address.save
```
Comment (kalpana19sharma) [inline app/controllers/addresses_controller.rb]: Don't call `save` after `create`

File: app/controllers/addresses_controller.rb [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+  def address_params
+    params.require(:address).permit(:country, :state, :city, :address, :pincode)
+  end
+
+  def set_user
+    @user = User.find_by_id(current_user.id)
+  end
+
+  def set_address
+    @address = Address.find_by_id(params[:id])
```
Comment (kalpana19sharma) [inline app/controllers/addresses_controller.rb:49]: Use associations.

File: app/controllers/addresses_controller.rb [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    redirect_to user_addresses_path
+  end
+
+  private
+  def address_params
+    params.require(:address).permit(:country, :state, :city, :address, :pincode)
+  end
+
+  def set_user
+    @user = User.find_by_id(current_user.id)
```
Comment (kalpana19sharma) [inline app/controllers/addresses_controller.rb]: Why are you executing same query again?

File: app/controllers/addresses_controller.rb (hunk 2 above)
Comment (kalpana19sharma) [reply app/controllers/addresses_controller.rb:49]: Handle the case if address is not found.

File: app/models/user.rb [hunk 4]
```diff
@@ -1,4 +1,5 @@
 class User < ApplicationRecord
+  has_many :addresses
```
Comment (kalpana19sharma) [inline app/models/user.rb]: Use dependent destroy

File: app/views/addresses/index.html.erb [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+        <th scope = "col">Country</th>
+        <th scope = "col">State</th>
+        <th scope = "col">City</th>
+        <th scope = "col">Address</th>
+        <th scope = "col">Pincode</th>
+        <th scope = "col">Options</th>
+      </tr>
+    </thead>
+    <tbody>
+      <% @address.each_with_index do |addr, index_no| %>
```
Comment (kalpana19sharma) [inline app/views/addresses/index.html.erb]: I think you can specify start value of index

File: config/routes.rb [hunk 6]
```diff
@@ -1,6 +1,8 @@
 Rails.application.routes.draw do
   root to: 'users#new'
-  resources :users
+  resources :users do
+    resources :addresses
```
Comment (kalpana19sharma) [inline config/routes.rb]: Are you using `show` action as well?

**Review by kalpana19sharma — CHANGES_REQUESTED** [review]

File: app/controllers/addresses_controller.rb (hunk 2 above)
Comment (anim-13) [reply app/controllers/addresses_controller.rb:49]: Mam, I had handled the case in view i.e. if address not found. The result will be printed
No Address Available.

File: config/routes.rb (hunk 6 above)
Comment (anim-13) [reply config/routes.rb]: No mam

File: app/views/addresses/index.html.erb (hunk 5 above)
Comment (anim-13) [reply app/views/addresses/index.html.erb:16]: when i am trying to use:
@address.each_with_index(1) it is showing following error:
wrong number of arguments (given 1, expected 0)
that's why i didn't specify the value of index

File: app/controllers/addresses_controller.rb [hunk 7]
```diff
@@ -0,0 +1,50 @@
+class AddressesController < ApplicationController
+  before_action :set_user, :set_address
+
+  def index
+    @address = current_user.addresses
```
Comment (kalpana19sharma) [inline app/controllers/addresses_controller.rb]: Variable name should be plural

File: app/controllers/addresses_controller.rb [hunk 8]
```diff
@@ -0,0 +1,50 @@
+class AddressesController < ApplicationController
+  before_action :set_user, :set_address
```
Comment (kalpana19sharma) [inline app/controllers/addresses_controller.rb]: Do we want to run these filters on all of the actions?

File: app/controllers/addresses_controller.rb [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+    redirect_to user_addresses_path
+  end
+
+  private
+  def address_params
+    params.require(:address).permit(:country, :state, :city, :address, :pincode)
+  end
+
+  def set_user
+    @user = User.find_by_id(params[:id])
```
Comment (kalpana19sharma) [inline app/controllers/addresses_controller.rb]: How can you use `params[:id]` for both user and address? Ultimately this record should be current user so you can directly use `current_user` whenever you need. Do we really need this method?

**Review by kalpana19sharma — CHANGES_REQUESTED** [review]

File: app/controllers/addresses_controller.rb (hunk 8 above)
Comment (anim-13) [reply app/controllers/addresses_controller.rb]: No

File: app/controllers/addresses_controller.rb [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+  end
+
+  private
+  def address_params
+    params.require(:address).permit(:country, :state, :city, :address, :pincode)
+  end
+
+  def set_user
+    @user = User.find_by_id(current_user)
+  end
```
Comment (kalpana19sharma) [inline app/controllers/addresses_controller.rb]:
```suggestion
  def set_user
    @user = current_user
  end
```

Comment (kalpana19sharma) [reply app/controllers/addresses_controller.rb]: In fact you can directly use `current_user` instead of `@user`

**Review by kalpana19sharma — CHANGES_REQUESTED** [review]

**Review by kalpana19sharma — APPROVED** [review]

**Pull request merged**